### DISCUSSION TITLE
Icons not being overridden anymore

### DISCUSSION THREAD

**Discussion by luckymachi** [kyazdani42/nvim-web-devicons#356] [Q&A] [answered]
Hello everyone! In my config I have overridden some icons (either the icons themselves or the color) to fit my theme/preferences more.
<details>
<summary>Here is an example image of how it looked:</summary>

![image](https://github.com/nvim-tree/nvim-web-devicons/assets/77021922/17f29f88-d4ab-43ad-b3f9-4e20f0258887)

</details>

However, out of nowhere it just stopped working. Just as extra info in case it may be the cause, I'm using NixOS and I'm managing my plugins through home-manager, but I still got the overrides to work properly like I've shown in the picture, so I really have no idea what could've changed.

<details>
<summary>Just in case, here's my config for it</summary>

```lua
-- Web-Devicons
require("nvim-web-devicons").setup({
	-- your personnal icons can go here (to override)
	-- you can specify color or cterm_color instead of specifying both of them
	-- DevIcon will be appended to `name`
	override = {
		rs = {
			icon = "",
			color = "#ff9e64",
			name = "Rust",
		},
		lua = {
			icon = "󰢱",
			color = "#7DCFFF",
			name = "Lua",
		},
		py = {
			icon = "",
			color = "#7bd88f",
			name = "Python",
		},
		go = {
			icon = "",
			color = "#7dcfff",
			name = "Go",
		},
		cpp = {
			icon = "",
			color = "#7aa2f7",
			name = "CPP",
		},
		c = {
			icon = "",
			color = "#BB9AF7",
			name = "C",
		},
		tex = {
			icon = "",
			color = "#9AA5CE",
			name = "LaTeX",
		},
		vim = {
			icon = "",
			color = "#7BD88F",
			name = "Vim",
		},
		oil = {
			icon = "󰏇",
			color = "#24283b",
			name = "oil",
		},
		vue = {
			icon = "",
			color = "#7bd88f",
			name = "Vue",
		},
	},
	override_by_extension = {
		["norg"] = {
			icon = "",
			color = "#4878BE",
			name = "Neorg",
		},
		["oil"] = {
			icon = "󰏇",
			color = "#24283b",
			name = "oil",
		},
	},
	-- globally enable different highlight colors per icon (default to true)
	-- if set to false all icons will have the default icon's color
	color_icons = true,
})
```

</details>

Thanks in advance for your time. Again, this really happened from one day to another (like two days ago it was all working perfectly) and I didn't change my config, at most I might've just updated my system (which shouldn't have changed my settings unless the plugin got updated), so any help on what could be going on is greatly appreciated!

**Comment by alex-courtis** [comment]
Overridden icons not showing implies an issue with plugin ordering.

You'll need call nvim-web-devicons setup before any other plugin setups. Almost all users of the icons cache them for performance reasons.

Check you plugin manager's documentation for information on how to manage startup ordering.

**Reply by luckymachi** [reply]
Hey there! So I've been trying to look into it a bit more and since I don't use any plugin-managers (at least from what can be understood by managing it through nix-config), I'll have to take a more manual approach with this.

So, I was wondering if, by managing the startup ordering, do you mean have it be the very first plugin that's loaded into the config, or the very latest, similar to what some people do to ``nvim/after/plugins/...``? 

Thanks for your time btw :)

**Reply by alex-courtis** [reply]
Unfortunately I don't know how nix-config works with its plugins.

The "normal" approach is to call your plugin manager such as packer then setup the plugins afterwards.

See https://github.com/nvim-tree/nvim-tree.lua/wiki/Installation for some examples of plugin managers. You'd call `require'nvim-web-devicons'.setup {` after the plugin manager is complete.

**Reply by luckymachi** [reply]
Yeah, thanks! I installed Lazy into my config to see if perhaps that could help, but still nothing :(
```
require("lazy").setup({
	{
		"nvim-tree/nvim-web-devicons",
		lazy = false,
		priority = 1000,
		opts = {
			override = {
				rs = {
					icon = "",
					color = "#ff9e64",
					name = "Rust",
				},
				lua = {
					icon = "󰢱",
					color = "#7DCFFF",
					name = "Lua",
				},
				py = {
					icon = "",
					color = "#7bd88f",
					name = "Python",
				},
				go = {
					icon = "",
					color = "#7dcfff",
					name = "Go",
				},
				cpp = {
					icon = "",
					color = "#7aa2f7",
					name = "CPP",
				},
				c = {
					icon = "",
					color = "#BB9AF7",
					name = "C",
				},
				tex = {
					icon = "",
					color = "#9AA5CE",
					name = "LaTeX",
				},
				vim = {
					icon = "",
					color = "#7BD88F",
					name = "Vim",
				},
				vue = {
					icon = "",
					color = "#7bd88f",
					name = "Vue",
				},
			},
			override_by_extension = {
				["norg"] = {
					icon = "",
					color = "#4878BE",
					name = "Neorg",
				},
				["py"] = {
					icon = "",
					color = "#7bd88f",
					name = "Python",
				},
			},
			-- globally enable different highlight colors per icon (default to true)
			-- if set to false all icons will have the default icon's color
			color_icons = true,
		},
	},
        {
		-- Set lualine as statusline
		"nvim-lualine/lualine.nvim",
		-- See `:help lualine.txt`
		opts = {
			options = {
				icons_enabled = true,
			},
		},
		config = function(_, opts)
			require("lualine").setup(opts)
		end,
		dependencies = {
			"nvim-tree/nvim-web-devicons",
		},
	},
       ...
```
I thought that this would finally do the trick but nope, nothing :(

![image](https://github.com/nvim-tree/nvim-web-devicons/assets/77021922/4caf7acb-9da3-4c95-8a38-f16de1fffe0a)

**Reply by alex-courtis** [reply]
Lazy has nondeterministic startup order and won't help you here. See https://github.com/nvim-tree/nvim-tree.lua/wiki/Installation#lazy-loading

**Reply by luckymachi** [reply]
Hey! I eventually changed the lazy-order but I decided to take another approach: in the file where I call my setup, after everything's supposed to be loaded, I'm calling for a local override to see if it's being overridden, and for some reason it is not applying. Here's the code in question: https://github.com/Maxito7/nixos_dotfiles/blob/master/home-manager/programs/neovim/nvim/lua/core/init.lua

![image](https://github.com/nvim-tree/nvim-web-devicons/assets/77021922/ede550f9-86c7-457c-8316-e6357d788d4e)
Here's the result of the assert test as well. Again, it's all super weird since it at least *should* load in this case

**Reply by alex-courtis** [reply]
Please try the [nvim-tree clean room configuration](https://github.com/nvim-tree/nvim-tree.lua/wiki/Troubleshooting#clean-room-replication )to rule out any issues with your nvim configuration.

**Reply by luckymachi** [reply]
Hello there! So I tried it out as mentioned and built it with the config I use for web-devicons hoping that they would at least properly override with NvimTree (which I don't use but I assume follows a similar priority to stuff like Lualine and Oil), but it seem that not even that is working properly :(

![image](https://github.com/nvim-tree/nvim-web-devicons/assets/77021922/dd5731c0-73d2-4b80-87a0-c5e10e00b905)

```lua
vim.g.loaded_netrw = 1
vim.g.loaded_netrwPlugin = 1

vim.cmd([[set runtimepath=$VIMRUNTIME]])
vim.cmd([[set packpath=/tmp/nvt-min/site]])
local package_root = "/tmp/nvt-min/site/pack"
local install_path = package_root .. "/packer/start/packer.nvim"
local function load_plugins()
	require("packer").startup({
		{
			"wbthomason/packer.nvim",
			"nvim-tree/nvim-tree.lua",
			"nvim-tree/nvim-web-devicons",
			-- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
		},
		config = {
			package_root = package_root,
			compile_path = install_path .. "/plugin/packer_compiled.lua",
			display = { non_interactive = true },
		},
	})
end
if vim.fn.isdirectory(install_path) == 0 then
	print("Installing nvim-tree and dependencies.")
	vim.fn.system({ "git", "clone", "--depth=1", "https://github.com/wbthomason/packer.nvim", install_path })
end
load_plugins()
require("packer").sync()
vim.cmd([[autocmd User PackerComplete ++once echo "Ready!" | lua setup()]])
vim.opt.termguicolors = true
vim.opt.cursorline = true

-- MODIFY NVIM-TREE SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
_G.setup = function()
	require("nvim-tree").setup({})
	require("nvim-web-devicons").setup({
		override = {
			rs = {
				icon = " ",
				color = "#ff9e64",
				name = "Rust",
			},
			lua = {
				icon = "󰢱",
				color = "#7DCFFF",
				name = "Lua",
			},
			py = {
				icon = "",
				color = "#7bd88f",
				name = "Python",
			},
			go = {
				icon = "",
				color = "#7dcfff",
				name = "Go",
			},
			cpp = {
				icon = "",
				color = "#7aa2f7",
				name = "CPP",
			},
			c = {
				icon = "",
				color = "#BB9AF7",
				name = "C",
			},
			tex = {
				icon = "",
				color = "#9AA5CE",
				name = "LaTeX",
			},
			vim = {
				icon = "",
				color = "#7BD88F",
				name = "Vim",
			},
			vue = {
				icon = "",
				color = "#7bd88f",
				name = "Vue",
			},
		},
		default = false,
		color_icons = true,
	})
end

-- UNCOMMENT this block for diagnostics issues, substituting pattern and cmd as appropriate.
-- Requires diagnostics.enable = true in setup.
vim.api.nvim_create_autocmd("FileType", {
	pattern = "lua",
	callback = function()
		vim.lsp.start({ cmd = { "lua-language-server" } })
	end,
})
```
I'll attempt to get the diagnostic things working, but so far this is very very weird

**Reply by alex-courtis** [reply, marked as answer]
The issue is ordering - you need to call web-devicons setup before you call nvim-tree setup, lualine setup etc. e.g.
```lua
	require("nvim-web-devicons").setup({
		override = {
      ---
		},
		default = false,
		color_icons = true,
	})
	require("nvim-tree").setup({})
```

**Reply by luckymachi** [reply]
Thank you! It did work as you mentioned :)

![image](https://github.com/nvim-tree/nvim-web-devicons/assets/77021922/6aa38afe-02b2-42c2-8b08-f9b34647c4bd)

It's definitely either a setup or packet manager thing, I'll try to look more into it but at least I know that the overrides themselves aren't wrong. Thanks a ton for the help and Happy New Year! :)
